crates/bevy_pbr/src/render/mesh_bindings.rs

```diff
@@ -1,229 +1,351 @@
+}
 
-    /// Also includes the uniform for skinning
-    pub skinned: BindGroupLayout,
+impl MeshBindGroups {
+    pub fn new() -> Self {
+        MeshBindGroups::default()
+    }
+    /// Get a specific [`BindGroup`] that was previously added.
+    pub fn get(
+        &self,
+        skin: bool,
+        morph_id: Option<AssetId<Mesh>>,
+        motion_vectors: bool,
+    ) -> Option<&BindGroup> {
+        let variant = ActiveVariant::new(skin, morph_id.is_some(), motion_vectors);
+        if let Some(id) = morph_id {
+            self.distinct.get(&(id, variant))
+        } else {
+            self.shared.get(&variant)
+        }
+    }
+    fn insert(
+        &mut self,
+        skin: bool,
+        morph_id: Option<AssetId<Mesh>>,
+        motion_vectors: bool,
+        bind_group: BindGroup,
+    ) {
+        let variant = ActiveVariant::new(skin, morph_id.is_some(), motion_vectors);
 
-    /// Also includes the uniform and [`MorphAttributes`] for morph targets.
-    ///
-    /// [`MorphAttributes`]: bevy_render::mesh::morph::MorphAttributes
-    pub morphed: BindGroupLayout,
+        if let Some(id) = morph_id {
+            self.distinct.insert((id, variant), bind_group);
+        } else {
+            self.shared.insert(variant, bind_group);
+        }
+    }
+    pub fn clear(&mut self) {
+        self.shared.clear();
+        self.distinct.clear();
+    }
+    /// Clears `self` and returns a [`MeshBindGroupBuilder`].
+    pub fn bind_group_builder<'a>(
+        &'a mut self,
+        device: &'a RenderDevice,
+        model: BindingResource<'a>,
+        layouts: &'a MeshLayouts,
+    ) -> MeshBindGroupBuilder<'a> {
+        self.clear();
 
-    /// Also includes both uniforms for skinning and morph targets, also the
-    /// morph target [`MorphAttributes`] binding.
-    ///
-    /// [`MorphAttributes`]: bevy_render::mesh::morph::MorphAttributes
-    pub morphed_skinned: BindGroupLayout,
+        MeshBindGroupBuilder {
+            layouts,
+            device,
+            model,
+            bind_groups: self,
+        }
+    }
 }
 
+/// All possible [`BindGroupLayout`]s in bevy's default mesh shader (`mesh.wgsl`).
+#[derive(Clone)]
+pub struct MeshLayouts([BindGroupLayout; ActiveVariant::COUNT]);
+
 impl MeshLayouts {
     /// Prepare the layouts used by the default bevy [`Mesh`].
     ///
     /// [`Mesh`]: bevy_render::prelude::Mesh
-    pub fn new(render_device: &RenderDevice) -> Self {
-        MeshLayouts {
-            model_only: Self::model_only_layout(render_device),
-            skinned: Self::skinned_layout(render_device),
-            morphed: Self::morphed_layout(render_device),
-            morphed_skinned: Self::morphed_skinned_layout(render_device),
-        }
+    pub fn new(device: &RenderDevice) -> Self {
+        let layout = |bits| variant_layout(ActiveVariant::from_bits_truncate(bits), device);
+        MeshLayouts(array::from_fn(layout))
+    }
+    // TODO: Passing 3 bools is pretty bad
```

Comment:
Just pass an ActiveVariant instead? `.get(ActiveVariant::SKIN | ActiveVariant::MOTION_VECTORS)` reads fine to me.

Oh, you have that just below. Then, is this function needed?